clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -291,6 +303,45 @@ private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime,
         return batches;
     }
 
+    boolean completed() {
+        return produceFuture.completed();
+    }
+
+    public void await() throws InterruptedException {
+        produceFuture.await();
+        waitForPossibleSplittedBatches();
+        clearChildrenProducerBatch();
+    }
+
+    private synchronized void waitForPossibleSplittedBatches() throws InterruptedException {
```

Comment:
This is synchronized and will await on the produceFuture. `await()` is called by `awaitFlushCompletion()` which is called when a user calls `flush()`. I am concerned that a user can call `flush()` and end up effectively dead locking other operations on the ProducerBatch, as getChildrenProducerBatch and addChildrenProducerBatch will not be able to be called by other threads - my concern is that the sender thread may become deadlocked in splitAndReenqueue in this state.

Reply:
I see your point @lbradstreet if the `sender` thread is working on a split batch and the `producer#flush()` is called, both will be blocking each other since we need to synchronize the access to the `childrenProducerBatch` List.

In that case, we should use another mechanism to keep track of the created `ProducerBatch` when a split occurs.